### PR TITLE
Feature/improve activities import

### PR DESCRIPTION
**Pull request by yksolanki9**
Fixes https://github.com/ghostfolio/ghostfolio/issues/1497

UI Preview:
https://user-images.githubusercontent.com/39493102/209437707-86afb439-214e-42f7-88f7-28221073d4e8.mp4

### CODE REVIEW

File: libs/ui/src/lib/activities-table/activities-table.component.ts [hunk 1]
```diff
@@ -133,13 +154,39 @@ export class ActivitiesTableComponent implements OnChanges, OnDestroy {
...
+  }
+
   public onChangePage(page: PageEvent) {
     this.pageIndex = page.pageIndex;
 
     this.totalFees = this.getTotalFees();
     this.totalValue = this.getTotalValue();
   }
 
+  public onClickActivity(activity: Activity) {
```
Comment (dtslvr) [inline libs/ui/src/lib/activities-table/activities-table.component.ts:175]: Great! 👍🏻

File: libs/ui/src/lib/activities-table/activities-table.component.scss [hunk 2]
```diff
@@ -25,6 +25,10 @@
         }
       }
 
+      .mat-checkbox {
```
Comment (dtslvr) [inline libs/ui/src/lib/activities-table/activities-table.component.scss:28]: I would remove this block here

File: libs/ui/src/lib/activities-table/activities-table.component.html [hunk 3]
```diff
@@ -15,6 +15,24 @@
     matSortDirection="desc"
     [dataSource]="dataSource"
   >
+    <ng-container matColumnDef="select">
+      <th *matHeaderCellDef class="px-1" mat-header-cell>
+        <mat-checkbox
```
Comment (dtslvr) [inline libs/ui/src/lib/activities-table/activities-table.component.html:20]: Add `class="mt-2"`

File: libs/ui/src/lib/activities-table/activities-table.component.html [hunk 4]
```diff
@@ -15,6 +15,24 @@
...
+          [checked]="selectedRows.hasValue() && areAllRowsSelected()"
+          [indeterminate]="selectedRows.hasValue() && !areAllRowsSelected()"
+          (change)="$event ? toggleAllRows() : null"
+        ></mat-checkbox>
+      </th>
+      <td *matCellDef="let element" class="px-1" mat-cell>
+        <mat-checkbox
```
Comment (dtslvr) [inline libs/ui/src/lib/activities-table/activities-table.component.html:27]: Add `class="mt-2"`

File: apps/client/src/app/pages/portfolio/activities/import-activities-dialog/import-activities-dialog.component.ts [hunk 5]
```diff
@@ -39,7 +43,39 @@ export class ImportActivitiesDialog implements OnDestroy {
     this.dialogRef.close();
   }
 
-  public onImport() {
+  public async onImportActivities() {
+    try {
+      await this.importActivitiesService.importSelectedActivities(
```
Comment (dtslvr) [inline apps/client/src/app/pages/portfolio/activities/import-activities-dialog/import-activities-dialog.component.ts:48]: Can you add the following snack bar before actually importing?

```
this.snackBar.open('⏳ ' + $localize`Importing data...`);
```

File: apps/client/src/app/pages/portfolio/activities/import-activities-dialog/import-activities-dialog.component.ts [hunk 6]
```diff
@@ -39,7 +43,39 @@ export class ImportActivitiesDialog implements OnDestroy {
...
+    }
+  }
+
+  public onReset() {
+    this.details = [];
+    this.errorMessages = [];
+    this.isFileSelected = false;
+  }
+
+  public onSelectFile() {
```
Comment (dtslvr) [inline apps/client/src/app/pages/portfolio/activities/import-activities-dialog/import-activities-dialog.component.ts:78]: Change the snack bar message below from 
```
'⏳ ' + $localize`Importing data...`
```
to
```
'⏳ ' + $localize`Validating data...`
```

**Review by dtslvr — CHANGES_REQUESTED** [review]
I have tried it and I added some comments. It already looks very good!

File: apps/client/src/app/services/import-activities.service.ts [hunk 7]
```diff
@@ -52,28 +55,56 @@ export class ImportActivitiesService {
...
         });
     });
   }
 
+  public importSelectedActivities(
+    selectedActivities: Activity[]
+  ): Promise<Activity[]> {
+    const importData: CreateOrderDto[] = [];
+    for (const activity of selectedActivities) {
+      importData.push({
```
Comment (dtslvr) [inline apps/client/src/app/services/import-activities.service.ts:94]: Extract to `convertToCreateOrderDto(aActivity: Activity)`

File: apps/client/src/app/pages/portfolio/activities/import-activities-dialog/import-activities-dialog.component.ts (hunk 5 above)
Comment (yksolanki9) [reply apps/client/src/app/pages/portfolio/activities/import-activities-dialog/import-activities-dialog.component.ts:48]: Done

File: apps/client/src/app/pages/portfolio/activities/import-activities-dialog/import-activities-dialog.component.ts (hunk 6 above)
Comment (yksolanki9) [reply apps/client/src/app/pages/portfolio/activities/import-activities-dialog/import-activities-dialog.component.ts:78]: Done

File: apps/client/src/app/pages/portfolio/activities/import-activities-dialog/import-activities-dialog.component.ts [hunk 8]
```diff
@@ -93,12 +128,11 @@ export class ImportActivitiesDialog implements OnDestroy {
             return;
           } else if (file.name.endsWith('.csv')) {
             try {
-              await this.importActivitiesService.importCsv({
+              this.activities = await this.importActivitiesService.importCsv({
                 fileContent,
-                userAccounts: this.data.user.accounts
+                userAccounts: this.data.user.accounts,
+                dryRun: true
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply apps/client/src/app/pages/portfolio/activities/import-activities-dialog/import-activities-dialog.component.ts:134]: Done

File: apps/client/src/app/services/import-activities.service.ts [hunk 9]
```diff
@@ -26,11 +27,13 @@ export class ImportActivitiesService {
 
   public async importCsv({
     fileContent,
-    userAccounts
+    userAccounts,
+    dryRun = false
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply apps/client/src/app/services/import-activities.service.ts:31]: Done

File: apps/client/src/app/services/import-activities.service.ts [hunk 10]
```diff
@@ -26,11 +27,13 @@ export class ImportActivitiesService {
...
   }: {
     fileContent: string;
     userAccounts: Account[];
-  }) {
+    dryRun?: boolean;
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply apps/client/src/app/services/import-activities.service.ts:35]: Done

File: apps/client/src/app/services/import-activities.service.ts [hunk 11]
```diff
@@ -52,28 +55,56 @@ export class ImportActivitiesService {
       });
     }
 
-    await this.importJson({ content: activities });
+    return await this.importJson({ content: activities, dryRun });
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply apps/client/src/app/services/import-activities.service.ts:58]: Didnt understand this?

[the remark this answers is not in the document]
Comment (yksolanki9) [reply apps/client/src/app/services/import-activities.service.ts:58]: It's already in alphabetical order

File: libs/ui/src/lib/activities-table/activities-table.component.html (hunk 3 above)
Comment (yksolanki9) [reply libs/ui/src/lib/activities-table/activities-table.component.html:20]: Done

File: libs/ui/src/lib/activities-table/activities-table.component.html (hunk 4 above)
Comment (yksolanki9) [reply libs/ui/src/lib/activities-table/activities-table.component.html:27]: Done

File: libs/ui/src/lib/activities-table/activities-table.component.scss (hunk 2 above)
Comment (yksolanki9) [reply libs/ui/src/lib/activities-table/activities-table.component.scss:28]: Done

File: apps/client/src/app/services/import-activities.service.ts (hunk 7 above)
Comment (yksolanki9) [reply apps/client/src/app/services/import-activities.service.ts:94]: Done

**Review by dtslvr — APPROVED** [review]
Great work, @yksolanki9! 😎

**Pull request merged**